lib/ansible/plugins/inventory/docker_swarm.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018, Stefan Heitmueller <stefan.heitmueller@gmx.com>
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: docker_swarm
+    plugin_type: inventory
+    authors:
+      - Stefan Heitmüller (stefan.heitmueller@gmx.com)
+    short_description: Ansible dynamic inventory plugin for Docker swarm nodes.
+    requirements:
+        - python >= 2.7
+        - Docker SDK for Python > 1.10.0
+    description:
+        - Reads inventories from the Docker swarm API.
+        - Uses a YAML configuration file docker_swarm.[yml|yaml].
+    options:
+        plugin:
+            description: The name of this plugin, it should always be set to 'docker_swarm' for this plugin to recognize it as it's own.
+            type: str
+            required: true
+            choices: docker_swarm
+        host:
+            description: Socket of a Docker swarm manager node (tcp,unix).
+            type: str
+            required: true
+        verbose_output:
+            description: Toggle to (not) include all available nodes metadata (e.g. Platform, Architecture, OS, EngineVersion)
+            type: bool
+            default: yes
+        tls:
+            description: Connect using TLS without verifying the authenticity of the Docker host server.
+            type: bool
+            default: no
+        tls_verify:
+            description: Connect using TLS without verifying the authenticity of the Docker host server.
+            type: bool
+            default: no
+        key_path:
+            description: Path to the client's TLS key file.
+            type: path
+        cacert_path:
+            description: Use a CA certificate when performing server verification by providing the path to a CA certificate file.
+            type: path
+        cert_path:
+            description: Path to the client's TLS certificate file.
+            type: path
+        tls_hostname:
+            description: When verifying the authenticity of the Docker Host server, provide the expected name of the server.
+            type: str
+'''
+
+EXAMPLES = '''
+# docker_swarm.yml
+plugin: docker_swarm
+host: tcp://my-docker-host:2375
+'''
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.inventory.group import to_safe_group_name as original_safe
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.release import __version__
+
+try:
+    import docker
+    HAS_DOCKER = True
+except ImportError:
+    HAS_DOCKER = False
+
+
+class InventoryModule(BaseInventoryPlugin):
```

Comment:
i recommend using 'compose' to allow users to dynamically create groups based on the data this plugin produces

Reply:
An example how this works:
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/doc_fragments/constructed.py (needs to be included via `extends_documentation_fragment:`)
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/aws_ec2.py#L489-L500 (needs to go to `_populate()`)
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/aws_ec2.py#L125-L150 (add something similar as an example, so people see how to generate groups from labels)

Reply:
Done, thanks for the hint.

@WojciechowskiPiotr : dynamic groups can be build now upon metadata ;)

Reply:
> Done, thanks for the hint.
> 
> @WojciechowskiPiotr : dynamic groups can be build now upon metadata ;)

I'm fine with that as long as it is clearly documented :)